contracts/access_control/lib.rs

```diff
@@ -0,0 +1,251 @@
+            self.priviledges.get((account, role)).is_some()
+        }
+
+        /// Terminates the contract.
+        ///
+        /// can only be called by the contract owner
+        #[ink(message, selector = 4)]
+        pub fn terminate(&mut self) -> Result<()> {
+            let caller = self.env().caller();
+            let this = self.env().account_id();
+            self.check_role(caller, Role::Owner(this))?;
+            self.env().terminate_contract(caller)
+        }
+
+        /// Result wrapped version of `has_role`
+        ///
+        /// Returns Error variant MissingRole(Role) if the account does not carry a role
+        #[ink(message, selector = 5)]
+        pub fn check_role(&self, account: AccountId, role: Role) -> Result<()> {
+            if !self.has_role(account, role) {
+                return Err(AccessControlError::MissingRole(role));
+            }
+            Ok(())
+        }
+
+        fn emit_event<EE: EmitEvent<Self>>(emitter: EE, event: Event) {
+            emitter.emit_event(event);
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+        use ink_lang as ink;
+
+        #[ink::test]
+        fn access_control() {
+            let accounts = ink_env::test::default_accounts::<ink_env::DefaultEnvironment>();
+
+            let alice = accounts.alice;
+            let bob = accounts.alice;
+            let charlie = accounts.charlie;
+
+            let contract_address = accounts.django;
+
+            // alice deploys the access control contract
+            ink_env::test::set_caller::<ink_env::DefaultEnvironment>(alice);
+            ink_env::test::set_callee::<ink_env::DefaultEnvironment>(contract_address);
+            ink_env::test::set_account_balance::<ink_env::DefaultEnvironment>(
+                contract_address,
+                100,
+            );
+            let mut access_control = AccessControl::new();
+
+            // alice should be admin
+            assert!(
+                access_control.has_role(alice, Role::Admin(contract_address)),
+                "deployer is not admin"
+            );
+
+            // alice should be owner
+            assert!(
+                access_control.has_role(alice, Role::Owner(contract_address)),
+                "deployer is not owner"
+            );
+
+            // alice grants admin rights to bob
+            assert!(
+                access_control
+                    .grant_role(bob, Role::Admin(contract_address))
+                    .is_ok(),
+                "Alice's grant_role call failed"
+            );
+
+            assert!(
+                access_control.has_role(bob, Role::Admin(contract_address)),
+                "Bob is not admin"
+            );
+
+            ink_env::test::set_caller::<ink_env::DefaultEnvironment>(bob);
+            ink_env::test::set_callee::<ink_env::DefaultEnvironment>(contract_address);
+
+            // bob grants admin rights to charlie
+            assert!(
```

Comment:
It seems like you check the same operation here, right?

Reply:
removed